Enhance security and implement cash journal functionality

Fabi911 opened the pull request:
- add Backend CRUD for Cash Journal
- add Frontend Income- and Expense Board

Fabi911 committed:
Enhance security and implement cash journal feature

Added GROUP1 role and corresponding permissions in SecurityConfig. Introduced cash journal functionality with new Booking and Type models, CashJournalController, service layer, and repository for data management.

Fabi911 committed:
Enhance security and implement cash journal feature

Added GROUP1 role and corresponding permissions in SecurityConfig. Introduced cash journal functionality with new Booking and Type models, CashJournalController, service layer, and repository for data management.

Fabi911 committed:
Enhance security and implement cash journal functionality

Added GROUP1 role and permissions in SecurityConfig for cash journal routes. Introduced CashJournal component, IncomeBoard and ExpenseBoard with new Booking type in frontend.

Fabi911 committed:
Add tests for CashJournal service and controller

Introduced unit tests for CashJournalService and integration tests for CashJournalController. These cover CRUD operations ensuring backend functionality is reliable and working as expected.

Fabi911 committed:
Refactor Booking tables for unified component

Deleted the ExpenseBoard component and combined functionality into a single BookingTable component. Renamed IncomeBoard to BookingTable and modified it to accept a type prop for differentiation between expense and income data.

Fabi911 committed:
Enhance loading state and adjust heading styles

Added MUI Skeleton components to improve the loading state in CashJournal, providing a better user experience. Also, fixed import paths and adjusted the heading levels in BookingTable for better visual hierarchy.

esgoet commented on backend/src/main/java/de/webdev/backend/cashjournal/service/CashJournalService.java:
```diff
@@ -0,0 +1,52 @@
+
+	public Booking updateBooking(String id, Booking booking) {
+		return cashJournalRepository.findById(id)
+				.map(existingBooking -> {
```
not sure why you map here if you don't use existingBooking anywhere? could just use `@With` annotation, for example, and then return what you get back with `findById.withDate(booking.date()).withDescription(...` and so on

esgoet commented on backend/src/main/java/de/webdev/backend/security/AppuserRole.java:
```diff
@@ -2,5 +2,7 @@
 
 public enum AppuserRole {
```
not sure how you have it in the rest of your code, but maybe AppUser instead of Appuser?
```suggestion
public enum AppUserRole {
```

esgoet commented on backend/src/main/java/de/webdev/backend/security/config/SecurityConfig.java:
```diff
@@ -36,10 +37,20 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
 								.requestMatchers(HttpMethod.GET, apiMembers).hasRole(AppuserRole.ADMIN.name())
 								.requestMatchers(HttpMethod.POST, apiMembers).hasRole(AppuserRole.ADMIN.name())
 								.requestMatchers(HttpMethod.PUT, apiMembers).hasRole(AppuserRole.ADMIN.name())
+								.requestMatchers(HttpMethod.DELETE, apiMembers).hasRole(AppuserRole.GROUP1.name())
```
if it's the same for all types of requests, you could just not specify it and only write it once? but obv doesn't hurt, and if you do plan to differentiate between different http request methods, then this makes sense
```suggestion
								.requestMatchers(apiMembers).hasRole(AppuserRole.GROUP1.name())
```

esgoet commented on backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:
```diff
@@ -0,0 +1,132 @@
+				.andExpect(MockMvcResultMatchers.status().isOk())
+				.andExpect(MockMvcResultMatchers.content().json("""
+						{
+						    "date": "2023-01-22",
```
would this not also return the id?

esgoet replied:
also maybe test for when cash journal by id does not exist?

esgoet commented on backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:
```diff
@@ -0,0 +1,132 @@
+						"""))
+				.andExpect(MockMvcResultMatchers.status().isOk())
+				.andExpect(MockMvcResultMatchers.content().json("""
+						{
```
id?

esgoet commented on backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:
```diff
@@ -0,0 +1,132 @@
+						}
+						"""));
+		mockMvc.perform(MockMvcRequestBuilders.delete("/api/cash-journal/1"))
+				.andExpect(MockMvcResultMatchers.status().isOk());
```
could also test here now that another getAll request is now just returning an empty list

esgoet commented on backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	@Test
+	@WithMockUser(roles = "ADMIN")
```
```suggestion
	@WithMockUser(roles = "ADMIN")
	@DirtiesContext
```

esgoet commented on backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	@Test
+	@WithMockUser(roles = "ADMIN")
```
```suggestion
	@WithMockUser(roles = "ADMIN")
	@DirtiesContext
```

esgoet commented on backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	@Test
+	@WithMockUser(roles = "ADMIN")
```
```suggestion
	@WithMockUser(roles = "ADMIN")
	@DirtiesContext
```

esgoet commented on backend/src/test/java/de/webdev/backend/cashjournal/service/CashJournalServiceTest.java:
the goal here is to test the CashJournalService methods, so you need to create an instance of CashJournalService and then call those methods in then //When section

esgoet commented on backend/src/test/java/de/webdev/backend/cashjournal/service/CashJournalServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+
+class CashJournalServiceTest {
+
+	private final CashJournalRepository cashJournalRepository=mock(CashJournalRepository.class);
```
```suggestion
	private final CashJournalRepository cashJournalRepository=mock(CashJournalRepository.class);
	private final CashJournalService cashJournalService = new CashJournalService(cashJournalRepository);
```

esgoet reviewed (commented):
Some @DirtiesContext missing in ControllerTest, and your CashJournalServiceTest needs to be updated, otherwise just optional comments

Fabi911 committed:
Refactor service and controller tests for CashJournal

Introduce the CashJournalService in tests and refactor methods to use it. Updated service and controller to streamline booking updates and added @DirtiesContext to relevant tests to ensure proper application context reloading.
